### PR TITLE
test(lwq): polish and test gym_vector_env_manager

### PR DESCRIPTION
**Pull request by Hcnaeg**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ class GymVectorEnvManager(BaseEnvManager):
         Create an GymVectorEnvManager to manage multiple environments.
         Each Environment is run by a respective subprocess.
     Interfaces:
-        seed, ready_obs, step, reset
+        seed, ready_obs, step, reset, close
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:21]: set `self._closed = False` when create `self.env_manager`

File: ding/envs/env_manager/tests/conftest.py [hunk 2]
```diff
@@ -165,6 +169,17 @@ def get_subprecess_manager_cfg(env_num=4):
     }
     return EasyDict(manager_cfg)
 
+def get_gym_vector_manager_cfg(env_num=4):
+    manager_cfg = {
+        'env_cfg': [{
+            'name': 'name{}'.format(i),
+        } for i in range(env_num)],
+        'episode_num': 2,
```
Comment (PaParaZz1) [inline ding/envs/env_manager/tests/conftest.py:177]: use and test valid arguments of gym vector manager, such as `shared_memory`

File: ding/envs/env_manager/tests/test_gym_vector_env_manager.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+from gym.vector.async_vector_env import AsyncState
+
+
+@pytest.mark.unittest
+class TestGymVectorEnvManager:
+    def test_naive(self, setup_gym_vector_manager_cfg):
+        env_fn = setup_gym_vector_manager_cfg.pop('env_fn')
+        env_manager = GymVectorEnvManager(env_fn, setup_gym_vector_manager_cfg)
+        env_manager.seed([314 for _ in range(env_manager.env_num)])
+        assert env_manager._closed
```
Comment (PaParaZz1) [inline ding/envs/env_manager/tests/test_gym_vector_env_manager.py:18]: assert not env_manager._closed

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/envs/env_manager/gym_vector_env_manager.py (hunk 1 above)
Comment (Hcnaeg) [reply ding/envs/env_manager/gym_vector_env_manager.py:21]: is it ok to set self._closed = False at the beginning of reset method?

Comment (PaParaZz1) [reply ding/envs/env_manager/gym_vector_env_manager.py:21]: No, you can refer to the source code of gym, relevant components like subprocess are already created when calling construction method.

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 4]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
+
+    def close(self) -> None:
+        """
+        Overview:
+            Release the environment resources
+            Since not calling super.__init__, no need to release BaseEnvManager's resources
+        """
+        if self._closed:
+            return
+        self._closed = True
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:118]: you also need to close env_ref and add related unittest

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 5]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
+                    if self._env_episode_count[i] >= self._episode_num:
+                        self._env_states[i] = EnvState.DONE
+
         return timestep_collate_result
 
     @property
     def ready_obs(self) -> Dict[int, Any]:
-        return {i: self.reset_observations[i] for i in range(len(self.reset_observations))}
+        return {
+            i: self.reset_observations[i]
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:101]: rename `reset_obseration`, which can result in confusion

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 6]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
-            self.final_eval_reward[i] += timestep_collate_result[i].reward
-            if timestep_collate_result[i].done:
-                timestep_collate_result[i].info['final_eval_reward'] = self.final_eval_reward[i]
-                self.final_eval_reward[i] = 0
+        for i in range(self.env_num):
+            if i in env_ids_given:
+                timestep_collate_result[i] = BaseEnvTimestep(
+                    timestep[0][i], timestep[1][i], timestep[2][i], timestep[3][i]
+                )
+                self._final_eval_reward[i] += timestep_collate_result[i].reward
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:88]: you don't update `ready_obs` when step method is called, which makes `ready_obs` always the result of reset

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 7]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
     def step(self, actions: Dict[int, Any]) -> Dict[int, namedtuple]:
+        assert isinstance(actions, Dict), type(actions)
+
+        env_ids_given = list(actions.keys())
+        for env_id in range(self.env_num):
+            if env_id not in actions.keys():
+                actions[env_id] = self._env_ref.random_action()
+
+        # actions should be sorted by keys
+        actions = dict(sorted(actions.items()))
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:72]: why these actions must be sorted

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 8]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
         elem = list(actions.values())[0]
-        if isinstance(elem, np.ndarray) and elem.shape == (1, ):
-            actions = [v.item() for k, v in actions.items()]
-        timestep = self.env_manager.step(actions)
+
+        if isinstance(elem, np.ndarray):
+            if elem.shape == (1, ):
+                actions = [v.item() for k, v in actions.items()]
+            else:
+                actions = list(actions.values())
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:80]: you ignore the case that `elem` is not the instance of `np.ndarray`

File: ding/envs/env_manager/gym_vector_env_manager.py [hunk 9]
```diff
@@ -32,39 +33,88 @@ def __init__(self, env_fn: List[Callable], cfg: EasyDict) -> None:
...
+                timestep_collate_result[i] = BaseEnvTimestep(
+                    timestep[0][i], timestep[1][i], timestep[2][i], timestep[3][i]
+                )
+                self._final_eval_reward[i] += timestep_collate_result[i].reward
+                if timestep_collate_result[i].done:
+                    timestep_collate_result[i].info['final_eval_reward'] = self._final_eval_reward[i]
+                    self._final_eval_reward[i] = 0
+                    self._env_episode_count[i] += 1
+                    if self._env_episode_count[i] >= self._episode_num:
+                        self._env_states[i] = EnvState.DONE
```
Comment (PaParaZz1) [inline ding/envs/env_manager/gym_vector_env_manager.py:94]: if not, we should set env_state; After that, if all the env is in reset state, we should call `self.reset` to reset all the env

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**